### PR TITLE
Adding tangent+color+texcoord1 support to vertices

### PR DESCRIPTION
**Pull request by JCash**
Currently, the splitting of Collada models is missing.

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java [hunk 1]
```diff
@@ -361,79 +363,194 @@ public static ArrayList<String> loadMaterialNames(Scene scene) {
...
+            if (newMesh == null) {
+                oldToNewIndex = new HashMap<Integer, Integer>();
+                newIndices = new ArrayList<Integer>();
+
+                newMesh = new Mesh();
+                newMesh.material = inMesh.material;
+                newMesh.name = String.format("%s_%d", inMesh.name, outMeshes.size());
+
+                newMesh.texCoords0NumComponents = inMesh.texCoords0NumComponents;
+                newMesh.texCoords1NumComponents = inMesh.texCoords1NumComponents;
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java:447]: The mechanic is straightforward.
For each new mesh we copy some values, and allocate arrays of max size.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java [hunk 2]
```diff
@@ -361,79 +363,194 @@ public static ArrayList<String> loadMaterialNames(Scene scene) {
...
+
+            int newIndex0 = oldToNewIndex.getOrDefault(index0, -1);
+            int newIndex1 = oldToNewIndex.getOrDefault(index1, -1);
+            int newIndex2 = oldToNewIndex.getOrDefault(index2, -1);
+
+            if (newIndex0 == -1) {
+                newIndex0 = vcount++;
+                oldToNewIndex.put(index0, newIndex0);
+                copyVertex(inMesh, index0, newMesh, newIndex0);
+            }
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java:479]: If the index didn't exist in the look up for this new mesh, then we need to add this vertex to the mesh, and add the new index to the table.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java [hunk 3]
```diff
@@ -361,79 +363,194 @@ public static ArrayList<String> loadMaterialNames(Scene scene) {
...
+                if (newMesh.colors != null)
+                    newMesh.colors = Arrays.copyOf(newMesh.colors, vcount * 4);
+                if (newMesh.weights != null)
+                    newMesh.weights = Arrays.copyOf(newMesh.weights, vcount * 4);
+                if (newMesh.bones != null)
+                    newMesh.bones = Arrays.copyOf(newMesh.bones, vcount * 4);
+                if (newMesh.texCoords0 != null)
+                    newMesh.texCoords0 = Arrays.copyOf(newMesh.texCoords0, vcount * newMesh.texCoords0NumComponents);
+                if (newMesh.texCoords1 != null)
+                    newMesh.texCoords1 = Arrays.copyOf(newMesh.texCoords1, vcount * newMesh.texCoords1NumComponents);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java:522]: Shrink the arrays to their actual size.

File: engine/gamesys/src/gamesys/components/comp_model.cpp [hunk 4]
```diff
@@ -139,9 +140,14 @@ namespace dmGameSystem
         dmGraphics::VertexElement ve[] =
         {
                 {"position", 0, 3, dmGraphics::TYPE_FLOAT, false},
-                {"texcoord0", 1, 2, dmGraphics::TYPE_FLOAT, false},
-                {"normal", 2, 3, dmGraphics::TYPE_FLOAT, false},
+                {"normal", 1, 3, dmGraphics::TYPE_FLOAT, false},
+                {"tangent", 2, 3, dmGraphics::TYPE_FLOAT, false},
+                {"color", 3, 4, dmGraphics::TYPE_FLOAT, false},
+                {"texcoord0", 4, 2, dmGraphics::TYPE_FLOAT, false},
+                {"texcoord1", 5, 2, dmGraphics::TYPE_FLOAT, false},
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_model.cpp:147]: In the future we might want specific declarations depending on the actual data coming in. Here _all_ models have the same attributes.

File: engine/gamesys/src/gamesys/resources/res_model.cpp [hunk 5]
```diff
@@ -78,137 +63,73 @@ namespace dmGameSystem
...
-        dmGraphics::Type index_element_type = dmGraphics::TYPE_UNSIGNED_SHORT;
-
-        void* index_buffer = 0;
+            // We've moved the runtime check to the build step instead
+            if (!support_32bit_indices)
+            {
+                dmLogError("The platform doesn't support 32 bit index buffers. See the setting 'model.split_large_meshes'");
+                return buffers;
+            }
+        }
```
Comment (JCash) [inline engine/gamesys/src/gamesys/resources/res_model.cpp:126]: Output an error and return.
The loading doesn't fail, the model will be just not be there.

File: engine/rig/src/dmsdk/rig/rig.h [hunk 6]
```diff
@@ -98,14 +98,12 @@ namespace dmRig
...
-        float v;
-        float nx;
-        float ny;
-        float nz;
+        float pos[3];
+        float normal[3];
+        float tangent[3];
+        float color[4];
+        float uv0[2];
+        float uv1[2];
```
Comment (JCash) [inline engine/rig/src/dmsdk/rig/rig.h:106]: A bit easier to maintain and loop over arrays

File: engine/rig/src/rig.cpp [hunk 7]
```diff
@@ -750,84 +751,115 @@ namespace dmRig
...
+                if (has_tangents)
+                {
+                    Vector3 tangent_in(tangents_in[i*3+0], tangents_in[i*3+1], tangents_in[i*3+2]);
+                    tangent = normal_matrix * tangent_in;
+                    if (lengthSqr(tangent) > 0.0f) {
+                        normalize(tangent);
+                    }
+                    *tangents_buffer++ = tangent[0];
+                    *tangents_buffer++ = tangent[1];
+                    *tangents_buffer++ = tangent[2];
```
Comment (JCash) [inline engine/rig/src/rig.cpp:793]: We check if we have tangents, and if so we transform those as well.

File: engine/rig/src/rig.cpp [hunk 8]
```diff
@@ -970,61 +1038,28 @@ namespace dmRig
...
-#ifdef __EMSCRIPTEN__
-#ifdef NDEBUG
-        #error "DEF-3610 - Can't compile with NDEBUG since dmRig::GenerateVertexData currently depends on assert() to work on Emscripten builds."
-#else
-        assert(vertex_data_out != 0x0);
-#endif
-#endif
-
-        return vertex_data_out;
+        return WriteVertexData(mesh, positions_buffer, normals_buffer, tangents_buffer, vertex_data_out);
```
Comment (JCash) [inline engine/rig/src/rig.cpp:1062]: We _really_ want skinning on the GPU so we can deprecate this CPU code path.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ColladaUtil.java [hunk 9]
```diff
@@ -571,7 +571,56 @@ private static XMLNode getFirstNodeWithGeoemtry(Collection<XMLVisualScene> scene
...
+            floatArray[i] = (float) a.get(i);
+        }
+        return floatArray;
+    }
+
+    private static int[] toIntArray(List<Integer> a) {
+        return a.stream().mapToInt(x -> x).toArray();
+    }
+
+    private static ModelImporter.Mesh createModelImporterMesh(List<Float> position_list,
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ColladaUtil.java:586]: It was easier to convert a collada mesh into a ModelImporter.Mesh, and let that class do the splitting.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ColladaUtil.java [hunk 10]
```diff
@@ -777,60 +825,81 @@ public boolean equals(Object o) {
...
-        if(shared_vertex_indices.size() <= 65536)
-        {
-            // if we only need 16-bit indices, use this primarily. Less data to upload to GPU and ES2.0 core functionality.
-            indices_format = Rig.IndexBufferFormat.INDEXBUFFER_FORMAT_16;
-            indices_bytes = ByteBuffer.allocateDirect(mesh_index_list.size() * 2);
-            indices_bytes.order(ByteOrder.LITTLE_ENDIAN).asShortBuffer();
-            for (int i = 0; i < mesh_index_list.size();) {
-                indices_bytes.putShort(mesh_index_list.get(i++).shortValue());
+        for (int index : mesh_index_list) {
+            MeshVertexIndex ci = shared_vertex_indices.get(index);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ColladaUtil.java:854]: Our previous struct was optimized for reconstructing meshes at runtime.
But we want them explicitly laid out in their lists.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ColladaUtil.java [hunk 11]
```diff
@@ -777,60 +825,81 @@ public boolean equals(Object o) {
...
+
+        if (splitMeshes && vertex_count >= 65536) {
+            ModelUtil.splitMesh(miMesh, allMeshes);
+        } else {
+            allMeshes.add(miMesh);
+        }
+
+        for (ModelImporter.Mesh newMesh : allMeshes) {
+            ArrayList<String> materials = new ArrayList<String>();
+            modelBuilder.addMeshes(ModelUtil.loadMesh(newMesh, materials));
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ColladaUtil.java:900]: Now, we always convert it into a ModelImporter.Mesh, and reuse that code.

**Review by britzl — APPROVED** [review]
Big one! Looks good as far as I can tell.

File: engine/rig/src/rig.cpp (hunk 8 above)
Comment (Jhonnyg) [reply engine/rig/src/rig.cpp:1062]: +inf on that.. ;(

**Review by Jhonnyg — APPROVED** [review]
It looks great! Eventually we might want to specify vertex declarations in the materials so you can minimize static geometry that doesn't necessarily need all the streams we've added, but let's go with this for now!

**Comment by Jhonnyg** [comment]
Hm, I tested an example gltf with the latest build, and something seems off with the UVs:
Screenshot 2022-08-30 at 23 54 53

This is from blender, same file with no changes:
Screenshot 2022-08-30 at 23 55 08

**Comment by JCash** [comment]
> I think this is the reason, the gltf UV map is not set in 0..1 range but in 1..2 range originally

Not sure where the [0,1] limit comes from, perhaps some UV wrap in the sampler?
But this is not due to this PR right?
I'm thinking I'll merge this in anycase to the model-importer-gltf branch

**Pull request merged**